corehq/blobs/migrate_metadata.py

```diff
@@ -0,0 +1,323 @@
+            return doc_type.get_db().dbname
+
+        for key, types in groupby(sorted(self.couch_types, key=db_key), key=db_key):
+            slug = "%s-%s" % (self.slug, key)
+            yield mod.Migrator(slug, list(types), couch_migrator)
+
+        for rex in self.sql_reindexers:
+            slug = "%s-%s" % (self.slug, rex.model_class.__name__)
+            yield SqlMigrator(slug, rex(), BlobMetaMigrator)
+
+    def migrate(self, filename, *args, **kw):
+        def filen(n):
+            return None if filename is None else "{}.{}".format(filename, n)
+        migrated = 0
+        skipped = 0
+        for n, item in enumerate(self.iter_migrators()):
+            one_migrated, one_skipped = item.migrate(filen(n), *args, **kw)
+            migrated += one_migrated
+            skipped += one_skipped
+            print("\n")
+        if not skipped:
+            BlobMigrationState.objects.get_or_create(slug=self.slug)[0].save()
+        return migrated, skipped
+
+
+def make_migrators(mod):
+    # defer class definitions to work around circular import
+
+    class BlobMetaMigrator(mod.BaseDocMigrator):
+        """Migrate blob metadata to BlobMeta model"""
+
+        def __init__(self, *args, **kw):
+            super(BlobMetaMigrator, self).__init__(*args, **kw)
+            self.total_blobs = 0
+
+        def _backup_doc(self, doc):
+            pass
+
+        def _do_migration(self, doc):
+            if not doc.get("external_blobs"):
+                return True
+            type_code = self.get_type_code(doc)
+            obj = self.blob_helper(doc, self.couchdb, type_code)
+            db = get_db_alias_for_partitioned_doc(doc["_id"])
+            with connections[db].cursor() as cursor:
+                for name, meta in six.iteritems(obj.blobs):
+                    if meta.blobmeta_id is not None:
+                        # blobmeta already saved
+                        continue
+                    if obj.domain is None:
+                        print("Unknown domain: {!r}".format(obj))
+                    cursor.execute("""
+                        INSERT INTO blobs_blobmeta (
+                            domain,
+                            type_code,
+                            parent_id,
+                            name,
+                            key,
+                            content_type,
+                            content_length,
+                            created_on
+                        ) VALUES (%s, %s, %s, %s, %s, %s, %s, CLOCK_TIMESTAMP())
+                        ON CONFLICT (key) DO NOTHING
+                    """, params=[
+                        (UNKNOWN_DOMAIN if obj.domain is None else obj.domain),
+                        type_code,
+                        doc["_id"],
+                        name,
+                        meta.key,
+                        meta.content_type,
+                        meta.content_length or 0,
+                    ])
```

Comment:
I used a raw query here because django does not support the `ON CONFLICT` clause. This is necessary because a resumed migration will try to insert metadata for blobs that have already been processed previously. This should have better performance characteristics than doing a lookup for each `BlobMeta` record before inserting it.